# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1391027315 18000
#      Wed Jan 29 15:28:35 2014 -0500
# Node ID 1e148ecdeedddebbc532a44c9f73600263f8fb91
# Parent  4c064d3d27509492df7c0bac9401f1de6603936e
# Parent  a9cec5517d9255fa7ea35cd9d41384f434ab83ae
maint: Periodic merge of gui-release to default.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -2714,19 +2714,24 @@ if test $build_gui = yes; then
     ## Retrieve Qt compilation and linker flags
     QT_CPPFLAGS="`$PKG_CONFIG --cflags-only-I QtCore QtGui QtNetwork`"
     QT_LDFLAGS="`$PKG_CONFIG --libs-only-L QtCore QtGui QtNetwork`"
     QT_LIBS="`$PKG_CONFIG --libs-only-l QtCore QtGui QtNetwork`"
 
     case $host_os in
       *darwin*)
         ## Qt might be installed in framework
-        QT_LDFLAGS="`$PKG_CONFIG --libs-only-other QtCore QtGui QtNetwork | tr ' ' '\n' | $GREP -e '-F' | uniq | tr '\n' ' '`"
-        QT_LIBS="`$PKG_CONFIG --libs-only-other QtCore QtGui QtNetwork | tr ' ' '\n' | $GREP -v -e '-F' | uniq | tr '\n' ' '`"
-        AM_CONDITIONAL([AMCOND_LINK_ALL_DEPS], [test $link_all_deps = yes || test -n $QT_LDFLAGS])
+        if test -z "$QT_LIBS"; then
+          QT_LDFLAGS="`$PKG_CONFIG --libs-only-other QtCore QtGui QtNetwork | tr ' ' '\n' | $GREP -e '-F' | uniq | tr '\n' ' '`"
+          QT_LIBS="`$PKG_CONFIG --libs-only-other QtCore QtGui QtNetwork | tr ' ' '\n' | $GREP -v -e '-F' | uniq | tr '\n' ' '`"
+          ## Enabling link_all_deps works around libtool's imperfect handling
+          ## of the -F flag
+          AM_CONDITIONAL([AMCOND_LINK_ALL_DEPS],
+                         [test $link_all_deps = yes || test -n "$QT_LDFLAGS"])
+        fi
       ;;
     esac
 
     ## Check for Qt4
     if ! `$PKG_CONFIG --atleast-version=4.0.0 QtCore`; then
       warn_gui="Qt >= 4.0.0 not found -- disabling GUI"
       build_gui=no
     fi
diff --git a/doc/interpreter/Makefile.am b/doc/interpreter/Makefile.am
--- a/doc/interpreter/Makefile.am
+++ b/doc/interpreter/Makefile.am
@@ -54,18 +54,18 @@ EXAMPLE_FILES = \
   $(top_srcdir)/examples/@polynomial/polynomial.m \
   $(top_srcdir)/examples/@polynomial/polynomial_superiorto.m \
   $(top_srcdir)/examples/@polynomial/polyval.m \
   $(top_srcdir)/examples/@polynomial/set.m \
   $(top_srcdir)/examples/@polynomial/subsref.m \
   $(top_srcdir)/examples/addtwomatrices.cc \
   $(top_srcdir)/examples/celldemo.cc \
   $(top_srcdir)/examples/embedded.cc \
-  $(top_srcdir)/examples/fortdemo.cc \
-  $(top_srcdir)/examples/fortsub.f \
+  $(top_srcdir)/examples/fortrandemo.cc \
+  $(top_srcdir)/examples/fortransub.f \
   $(top_srcdir)/examples/funcdemo.cc \
   $(top_srcdir)/examples/globaldemo.cc \
   $(top_srcdir)/examples/helloworld.cc \
   $(top_srcdir)/examples/mycell.c \
   $(top_srcdir)/examples/myfeval.c \
   $(top_srcdir)/examples/myfunc.c \
   $(top_srcdir)/examples/myhello.c \
   $(top_srcdir)/examples/mypow2.c \
diff --git a/doc/interpreter/debug.txi b/doc/interpreter/debug.txi
--- a/doc/interpreter/debug.txi
+++ b/doc/interpreter/debug.txi
@@ -39,136 +39,135 @@ debug mode.
 * Profiler Example::
 @end menu
 
 @node Entering Debug Mode
 @section Entering Debug Mode
 
 There are two basic means of interrupting the execution of an Octave
 script.  These are breakpoints (@pxref{Breakpoints}), discussed in the next
-section and interruption based on some condition.
+section, and interruption based on some condition.
 
 Octave supports three means to stop execution based on the values set in
-the functions @code{debug_on_interrupt}, @code{debug_on_warning} and
+the functions @code{debug_on_interrupt}, @code{debug_on_warning}, and
 @code{debug_on_error}.
 
 @DOCSTRING(debug_on_interrupt)
 
 @DOCSTRING(debug_on_warning)
 
 @DOCSTRING(debug_on_error)
 
 @node Leaving Debug Mode
 @section Leaving Debug Mode
 
-To leave the debug mode, use either @code{dbcont} 
-or @code{return}.
+Use either @code{dbcont} or @code{return} to leave the debug mode and
+continue the normal execution of the script.
 
 @DOCSTRING(dbcont)
 
-To quit debug mode and return directly to the prompt @code{dbquit}
-should be used instead
+To quit debug mode and return directly to the prompt without executing
+any additional code use @code{dbquit}.
 
 @DOCSTRING(dbquit)
 
 Finally, typing @code{exit} or @code{quit} at the debug prompt will
 result in Octave terminating normally.
 
 @node Breakpoints
 @section Breakpoints
 
-Breakpoints can be set in any Octave function, using the @code{dbstop}
+Breakpoints can be set in any m-file function by using the @code{dbstop}
 function.
 
 @DOCSTRING(dbstop)
 
 @noindent
 Breakpoints in class methods are also supported (e.g.,
-@code{dbstop("@@class/method")}).  However, breakpoints cannot be set in
-built-in functions (e.g., @code{sin}, etc.) or dynamically loaded function
+@code{dbstop ("@@class/method")}).  However, breakpoints cannot be set in
+built-in functions (e.g., @code{sin}, etc.) or dynamically loaded functions
 (i.e., oct-files).
 
-To set a breakpoint immediately on entering a function, the breakpoint
-should be set to line 1. The leading comment block will be ignored and
-the breakpoint will be set to the first executable statement in the
-function.  For example:
+To set a breakpoint immediately upon entering a function use line number 1, or
+omit the line number entirely and just give the function name.  When setting
+the breakpoint Octave will ignore the leading comment block, and the breakpoint
+will be set on the first executable statement in the function.  For example:
 
 @example
 @group
 dbstop ("asind", 1)
 @result{} 29
 @end group
 @end example
 
 @noindent
 Note that the return value of @code{29} means that the breakpoint was
 effectively set to line 29.  The status of breakpoints in a function can
-be queried with the @code{dbstatus} function.
+be queried with @code{dbstatus}.
 
 @DOCSTRING(dbstatus)
 
 @noindent
 Reusing the previous example, @code{dbstatus ("asind")} will return
 29.  The breakpoints listed can then be cleared with the @code{dbclear}
 function.
 
 @DOCSTRING(dbclear)
 
 @noindent
-These functions can be used together to clear all the breakpoints in a
-particular function.  For example:
-
-@example
-dbclear ("asind", dbstatus ("asind"));
-@end example
 
 A breakpoint may also be set in a subfunction.  For example, if a file contains
 the functions
 
 @example
 @group
 function y = func1 (x)
   y = func2 (x);
 endfunction
 function y = func2 (x)
   y = x + 1;
 endfunction
 @end group
 @end example
 
 @noindent
-then a breakpoint can be set at the start of the subfunction directly
-with
+then a breakpoint can be set at the start of the subfunction directly with
 
 @example
 @group
 dbstop (["func1", filemarker(), "func2"])
 @result{} 5
 @end group
 @end example
 
-Note that @code{filemarker} returns a character that marks the
-subfunctions from the file containing them.
+Note that @code{filemarker} returns the character that marks subfunctions from
+the file containing them.  Unless the default has been changed this character
+is @samp{>}.  Thus, a quicker and more normal way to set the breakpoint would
+be
+
+@example
+dbstop func1>func2
+@end example
 
 Another simple way of setting a breakpoint in an Octave script is the
 use of the @code{keyboard} function.
 
 @DOCSTRING(keyboard)
 
 @noindent
-The @code{keyboard} function is typically placed in a script at the
-point where the user desires that the execution be stopped.  It
-automatically sets the running script into the debug mode.
+The @code{keyboard} function is placed in a script at the point where the user
+desires that the execution be stopped.  It automatically sets the running
+script into the debug mode.
 
 @node Debug Mode
 @section Debug Mode
 
 There are three additional support functions that allow the user to
-interrogate where in the execution of a script Octave entered the debug
-mode and to print the code in the script surrounding the point where
+find out where in the execution of a script Octave entered the debug
+mode, and to print the code in the script surrounding the point where
 Octave entered debug mode.
 
 @DOCSTRING(dbwhere)
 
 @DOCSTRING(dbtype)
 
 @DOCSTRING(dblist)
 
@@ -177,24 +176,31 @@ currently active.
 
 @DOCSTRING(isdebugmode)
 
 Debug mode also allows single line stepping through a function using
 the command @code{dbstep}.
 
 @DOCSTRING(dbstep)
 
-When in debug mode the @key{RETURN} will execute the last entered command.
+When in debug mode the @key{RETURN} key will execute the last entered command.
 This is useful, for example, after hitting a breakpoint and entering
-@code{dbstep}.  After that one can advance line by line through the code
+@code{dbstep} once.  After that, one can advance line by line through the code
 with only a single key stroke.
 
 @node Call Stack
 @section Call Stack
 
+The function being debugged may be the leaf node of a series of function calls.
+After examining values in the current subroutine it may turn out that the
+problem occurred in earlier pieces of code.  Use @code{dbup} and @code{dbdown}
+to move up and down through the series of function calls to locate where
+variables first took on the wrong values.  @code{dbstack} shows the entire
+series of function calls and at what level debugging is currently taking place.
+
 @DOCSTRING(dbstack)
 
 @DOCSTRING(dbup)
 
 @DOCSTRING(dbdown)
 
 @node Profiling
 @section Profiling
diff --git a/doc/interpreter/external.txi b/doc/interpreter/external.txi
--- a/doc/interpreter/external.txi
+++ b/doc/interpreter/external.txi
@@ -1032,40 +1032,40 @@ for Fortran code that is short running a
 there is potentially an overhead in doing so.  However, if @w{@code{F77_FCN}}
 is used with code that calls @code{XSTOP}, Octave can generate a
 segmentation fault.
 
 An example of the inclusion of a Fortran function in an oct-file is
 given in the following example, where the C++ wrapper is
 
 @example
-@EXAMPLEFILE(fortdemo.cc)
+@EXAMPLEFILE(fortrandemo.cc)
 @end example
 
 @noindent
 and the Fortran function is
 
 @example
-@EXAMPLEFILE(fortsub.f)
+@EXAMPLEFILE(fortransub.f)
 @end example
 
 This example demonstrates most of the features needed to link to an
 external Fortran function, including passing arrays and strings, as well
 as exception handling.  Both the Fortran and C++ files need to be compiled
 in order for the example to work.
 
 @example
 @group
-mkoctfile fortdemo.cc fortsub.f
-[b, s] = fortdemo (1:3)
+mkoctfile fortrandemo.cc fortransub.f
+[b, s] = fortrandemo (1:3)
 @result{}
   b = 1.00000   0.50000   0.33333
   s = There are   3 values in the input vector
-[b, s] = fortdemo (0:3)
-error: fortdemo: fortsub: divide by zero
+[b, s] = fortrandemo (0:3)
+error: fortrandemo: fortransub: divide by zero
 @end group
 @end example
 
 @node Allocating Local Memory in Oct-Files
 @subsection Allocating Local Memory in Oct-Files
 
 Allocating memory within an oct-file might seem easy as the C++
 new/delete operators can be used.  However, in that case great care must be
@@ -1707,22 +1707,21 @@ can be see in the example below.
 @EXAMPLEFILE(myfeval.c)
 @end example
 
 If this code is in the file @file{myfeval.c}, and is compiled to
 @file{myfeval.mex}, then an example of its use is
 
 @example
 @group
-myfeval ("sin", 1)
 a = myfeval ("sin", 1)
-@result{} Hello, World!
-    I have 2 inputs and 1 outputs
-    I'm going to call the interpreter function sin
-    a =  0.84147
+@result{} Starting file myfeval.mex
+   I have 2 inputs and 1 outputs
+   I'm going to call the interpreter function sin
+   a =  0.84147
 @end group
 @end example
 
 Note that it is not possible to use function handles or inline functions
 within a mex-file.
 
 @c @node Application Programming Interface for Mex-Files
 @c @subsection Application Programming Interface for Mex-Files
@@ -1806,13 +1805,12 @@ which, again, is compiled and run as a s
 $ mkoctfile --link-stand-alone standalonebuiltin.cc -o standalonebuiltin
 $ ./standalonebuiltin 
 This is a matrix:
  11 12
  21 22
 
 This is the norm of the matrix:
 34.4952
-
+$
 @end group
 @end example
 
-
diff --git a/doc/interpreter/linalg.txi b/doc/interpreter/linalg.txi
--- a/doc/interpreter/linalg.txi
+++ b/doc/interpreter/linalg.txi
@@ -15,45 +15,54 @@
 @c You should have received a copy of the GNU General Public License
 @c along with Octave; see the file COPYING.  If not, see
 @c <http://www.gnu.org/licenses/>.
 
 @node Linear Algebra
 @chapter Linear Algebra
 @cindex linear algebra
 
-This chapter documents the linear algebra functions of Octave.
-Reference material for many of these functions may be found in
-Golub and Van Loan, @cite{Matrix Computations, 2nd Ed.}, Johns Hopkins,
-1989, and in the @cite{@sc{lapack} Users' Guide}, SIAM, 1992.
+This chapter documents the linear algebra functions provided in Octave.
+Reference material for many of these functions may be found in Golub and
+Van Loan, @cite{Matrix Computations, 2nd Ed.}, Johns Hopkins, 1989, and
+in the @cite{@sc{lapack} Users' Guide}, SIAM, 1992. The
+@cite{@sc{lapack} Users' Guide} is available at:
+@cite{http://www.netlib.org/lapack/lug/}
+
+A common text for engineering courses is G. Strang, @cite{Linear Algebra
+and Its Applications, 4th Edition}. It has become a widespread reference
+for linear algebra. An alternative is P. Lax @cite{Linear Algebra and
+Its Applications}, and also is a good choice. It claims to be suitable
+for high school students with substantial mathematical interests as well
+as first-year undergraduates.
 
 @menu
 * Techniques Used for Linear Algebra::
 * Basic Matrix Functions::
 * Matrix Factorizations::
 * Functions of a Matrix::
 * Specialized Solvers::
 @end menu
 
 @node Techniques Used for Linear Algebra
 @section Techniques Used for Linear Algebra
 @cindex linear algebra, techniques
 
-Octave includes a polymorphic solver, that selects an appropriate
-matrix factorization depending on the properties of the matrix itself. 
-Generally, the cost of determining the matrix type is small relative to 
-the cost of factorizing the matrix itself, but in any case the matrix 
-type is cached once it is calculated, so that it is not re-determined 
-each time it is used in a linear equation.
+Octave includes a polymorphic solver that selects an appropriate matrix
+factorization depending on the properties of the matrix itself.
+Generally, the cost of determining the matrix type is small relative to
+the cost of factorizing the matrix itself. In any case the matrix type
+is cached once it is calculated so that it is not re-determined each
+time it is used in a linear equation.
 
-The selection tree for how the linear equation is solve or a matrix
-inverse is form is given by
+The selection tree for how the linear equation is solved or a matrix
+inverse is formed is given by:
 
 @enumerate 1
-@item If the matrix is upper or lower triangular sparse a forward or
+@item If the matrix is upper or lower triangular sparse use a forward or
 backward substitution using the @sc{lapack} xTRTRS function, and goto 4.
 
 @c Permuted triangular matrices currently disabled in the code
 @c
 @c @item If the matrix is a upper triangular matrix with column permutations
 @c or lower triangular matrix with row permutations, perform a forward or
 @c backward substitution, and goto 5.
 
@@ -71,20 +80,20 @@ the @sc{lapack} xGELSD function.
 
 The user can force the type of the matrix with the @code{matrix_type}
 function.  This overcomes the cost of discovering the type of the matrix.
 However, it should be noted that identifying the type of the matrix incorrectly
 will lead to unpredictable results, and so @code{matrix_type} should be
 used with care.
 
 It should be noted that the test for whether a matrix is a candidate for
-Cholesky@tie{}factorization, performed above and by the @code{matrix_type}
-function, does not give a certainty that the matrix is
+Cholesky@tie{}factorization, performed above, and by the @code{matrix_type}
+function, does not make certain that the matrix is
 Hermitian.  However, the attempt to factorize the matrix will quickly
-flag a non-Hermitian matrix.
+detect a non-Hermitian matrix.
 
 @node Basic Matrix Functions
 @section Basic Matrix Functions
 @cindex matrix functions, basic
 
 @DOCSTRING(balance)
 
 @DOCSTRING(cond)
diff --git a/examples/Makefile.am b/examples/Makefile.am
--- a/examples/Makefile.am
+++ b/examples/Makefile.am
@@ -29,18 +29,18 @@ FCN_FILES =
 include @polynomial/module.mk
 include @FIRfilter/module.mk
 
 EXTRA_DIST += \
   $(FCN_FILES) \
   addtwomatrices.cc \
   celldemo.cc \
   embedded.cc \
-  fortdemo.cc \
-  fortsub.f \
+  fortrandemo.cc \
+  fortransub.f \
   funcdemo.cc \
   globaldemo.cc \
   helloworld.cc \
   make_int.cc \
   mex_demo.c \
   mycell.c \
   myfeval.c \
   myfevalf.f \
diff --git a/examples/fortdemo.cc b/examples/fortrandemo.cc
rename from examples/fortdemo.cc
rename to examples/fortrandemo.cc
--- a/examples/fortdemo.cc
+++ b/examples/fortrandemo.cc
@@ -1,35 +1,35 @@
 #include <octave/oct.h>
 #include <octave/f77-fcn.h>
 
 extern "C"
 {
   F77_RET_T
-  F77_FUNC (fortsub, FORTSUB)
+  F77_FUNC (fortransub, FORTSUB)
     (const int&, double*, F77_CHAR_ARG_DECL F77_CHAR_ARG_LEN_DECL);
 }
 
-DEFUN_DLD (fortdemo, args, , "Fortran Demo")
+DEFUN_DLD (fortrandemo, args, , "Fortran Demo")
 {
   octave_value_list retval;
   int nargin = args.length ();
 
   if (nargin != 1)
     print_usage ();
   else
     {
       NDArray a = args(0).array_value ();
       if (! error_state)
         {
           double *av = a.fortran_vec ();
           octave_idx_type na = a.numel ();
           OCTAVE_LOCAL_BUFFER (char, ctmp, 128);
 
-          F77_XFCN (fortsub, FORTSUB,
+          F77_XFCN (fortransub, FORTSUB,
                     (na, av, ctmp F77_CHAR_ARG_LEN (128)));
 
           retval(1) = std::string (ctmp);
           retval(0) = a;
         }
     }
   return retval;
 }
diff --git a/examples/fortsub.f b/examples/fortransub.f
rename from examples/fortsub.f
rename to examples/fortransub.f
--- a/examples/fortsub.f
+++ b/examples/fortransub.f
@@ -1,21 +1,21 @@
-      subroutine fortsub (n, a, s)
+      subroutine fortransub (n, a, s)
       implicit none
       character*(*) s
       real*8 a(*)
       integer*4 i, n, ioerr
       do i = 1, n
         if (a(i) .eq. 0d0) then
-          call xstopx ('fortsub: divide by zero')
+          call xstopx ('fortransub: divide by zero')
         else
           a(i) = 1d0 / a(i)
         endif
       enddo
       write (unit = s, fmt = '(a,i3,a,a)', iostat = ioerr)
      $       'There are ', n,
      $       ' values in the input vector', char(0)
       if (ioerr .ne. 0) then
-        call xstopx ('fortsub: error writing string')
+        call xstopx ('fortransub: error writing string')
       endif
       return
       end
 
diff --git a/examples/mex_demo.c b/examples/mex_demo.c
--- a/examples/mex_demo.c
+++ b/examples/mex_demo.c
@@ -5,23 +5,23 @@
 //
 //   octave_config_info ("ENABLE_DYNAMIC_LINKING")
 //
 // at the Octave prompt.  Support for dynamic linking is included if
 // this expression returns the string "yes".
 //
 // To compile this file, type the command
 //
-//   mkoctfile mex_demo.c
+//   mkoctfile --mex mex_demo.c
 //
 // from within Octave or from the shell prompt.  This will create a file
 // called mex_demo.mex that can be loaded by Octave.  To test the mex_demo.mex
 // file, start Octave and type the command
 //
-// [d] = mex_demo ("easy as", 1, 2, 3)
+// d = mex_demo ("easy as", 1, 2, 3)
 //
 // at the Octave prompt.  Octave should respond by printing
 //
 //   Hello, world!
 //   I have 4 inputs and 1 output
 //   d =  1.2346
 
 // Additional samples of code are in the examples directory of the Octave
diff --git a/examples/myfeval.c b/examples/myfeval.c
--- a/examples/myfeval.c
+++ b/examples/myfeval.c
@@ -1,23 +1,27 @@
 #include "mex.h"
 
 void
 mexFunction (int nlhs, mxArray* plhs[],
              int nrhs, const mxArray* prhs[])
 {
   char *str;
 
-  mexPrintf ("Hello, World!\n");
+  mexPrintf ("Starting file myfeval.mex\n");
 
   mexPrintf ("I have %d inputs and %d outputs\n", nrhs, nlhs);
 
   if (nrhs < 1 || ! mxIsString (prhs[0]))
     mexErrMsgTxt ("ARG1 must be a function name");
 
   str = mxArrayToString (prhs[0]);
 
   mexPrintf ("I'm going to call the function %s\n", str);
 
-  mexCallMATLAB (nlhs, plhs, nrhs-1, (mxArray*)prhs+1, str);
+  if (nlhs == 0)
+    nlhs = 1;  // Octave's automatic 'ans' variable
+
+  /* Cast prhs just to get rid of 'const' qualifier and stop compile warning */
+  mexCallMATLAB (nlhs, plhs, nrhs-1, (mxArray**)prhs+1, str);
 
   mxFree (str);
 }
diff --git a/libgui/languages/be_BY.ts b/libgui/languages/be_BY.ts
--- a/libgui/languages/be_BY.ts
+++ b/libgui/languages/be_BY.ts
@@ -893,17 +893,17 @@ Do you want to delete it anyway?</source
     </message>
     <message>
         <location line="+1"/>
         <source>Write %1 file</source>
         <translation>Запісаць %1-файл</translation>
     </message>
     <message>
         <location line="+71"/>
-        <source>Print To File ...</source>
+        <source>Print To File...</source>
         <translation>Друк у файл...</translation>
     </message>
     <message>
         <location line="+78"/>
         <source>%1 is a directory.
 Please choose a different file name.</source>
         <translation>%1 - гэта каталог.
 Калі ласка, пазначце іншую назву файла.</translation>
@@ -2060,17 +2060,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation>Шукаць файлы...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Новы файл</translation>
     </message>
     <message>
@@ -2106,16 +2106,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Перайсці да хатняга каталогу</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Шукаць каталог...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Шукаць файлы...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Новы файл...</translation>
     </message>
diff --git a/libgui/languages/de_DE.ts b/libgui/languages/de_DE.ts
--- a/libgui/languages/de_DE.ts
+++ b/libgui/languages/de_DE.ts
@@ -520,18 +520,18 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
-        <translation>Dateien suchen ...</translation>
+        <source>Find Files...</source>
+        <translation>Dateien suchen...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Neue Datei</translation>
     </message>
     <message>
         <location line="+3"/>
@@ -566,19 +566,24 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Wechsle zum Heimatverzeichnis</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Verzeichnis suchen...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translatorcomment>Dateien suchen...</translatorcomment>
-        <translation>Suche Dateien...</translation>
+        <translation>Dateien suchen...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Neue Datei...</translation>
     </message>
     <message>
         <location line="+3"/>
@@ -909,17 +914,17 @@ a:link { text-decoration: underline; col
     <message>
         <location line="+5"/>
         <source>Enter the start directory</source>
         <translation>Verzeichnis, in dem die Suche beginnt</translation>
     </message>
     <message>
         <location line="+3"/>
         <source>Browse...</source>
-        <translation>Suche ...</translation>
+        <translation>Suche...</translation>
     </message>
     <message>
         <location line="+1"/>
         <source>Browse for start directory</source>
         <translation>Suche Startverzeichnis</translation>
     </message>
     <message>
         <location line="+7"/>
@@ -1019,17 +1024,17 @@ a:link { text-decoration: underline; col
     <message>
         <location line="+17"/>
         <source>File contents</source>
         <translation>Dateiinhalt</translation>
     </message>
     <message>
         <location line="+105"/>
         <source>Searching...</source>
-        <translation>Suche ...</translation>
+        <translation>Suche...</translation>
     </message>
     <message>
         <location line="+23"/>
         <source>%1 match(es)</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+9"/>
diff --git a/libgui/languages/en_US.ts b/libgui/languages/en_US.ts
--- a/libgui/languages/en_US.ts
+++ b/libgui/languages/en_US.ts
@@ -503,17 +503,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
@@ -549,16 +549,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation type="unfinished"></translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation type="unfinished"></translation>
     </message>
diff --git a/libgui/languages/es_ES.ts b/libgui/languages/es_ES.ts
--- a/libgui/languages/es_ES.ts
+++ b/libgui/languages/es_ES.ts
@@ -503,18 +503,18 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
-        <translation>Buscar archivos ...</translation>
+        <source>Find Files...</source>
+        <translation>Buscar archivos...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Nuevo archivo</translation>
     </message>
     <message>
         <location line="+3"/>
@@ -549,16 +549,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Mostrar el directorio de inicio</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Buscar directorio...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Buscar archivos...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Nuevo archivo...</translation>
     </message>
diff --git a/libgui/languages/fr_FR.ts b/libgui/languages/fr_FR.ts
--- a/libgui/languages/fr_FR.ts
+++ b/libgui/languages/fr_FR.ts
@@ -519,17 +519,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation>Rechercher des fichiers...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Nouveau fichier</translation>
     </message>
     <message>
@@ -565,16 +565,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Aller au répertoire personnel</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Rechercher dans le répertoire...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Rechercher des fichiers...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Nouveau fichier...</translation>
     </message>
diff --git a/libgui/languages/nl_NL.ts b/libgui/languages/nl_NL.ts
--- a/libgui/languages/nl_NL.ts
+++ b/libgui/languages/nl_NL.ts
@@ -892,17 +892,17 @@ Wil je die toch verwijderen?</translatio
     </message>
     <message>
         <location line="+1"/>
         <source>Write %1 file</source>
         <translation>Bewaar %1 bestand</translation>
     </message>
     <message>
         <location line="+71"/>
-        <source>Print To File ...</source>
+        <source>Print To File...</source>
         <translation>Druk af naar bestand...</translation>
     </message>
     <message>
         <location line="+78"/>
         <source>%1 is a directory.
 Please choose a different file name.</source>
         <translation>%1 is een map.
 Kies s.v.p. een andere bestandsnaam.</translation>
@@ -2058,18 +2058,18 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
-        <translation>Zoek bestanden ...</translation>
+        <source>Find Files...</source>
+        <translation>Zoek bestanden...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Nieuw bestand</translation>
     </message>
     <message>
         <location line="+3"/>
@@ -2104,16 +2104,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Laat home map zien</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Zoek in map...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Zoek bestanden...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Nieuw bestand...</translation>
     </message>
@@ -2448,17 +2453,17 @@ a:link { text-decoration: underline; col
     <message>
         <location line="+5"/>
         <source>Enter the start directory</source>
         <translation>Geef startmap op</translation>
     </message>
     <message>
         <location line="+3"/>
         <source>Browse...</source>
-        <translation>Bladeren ...</translation>
+        <translation>Bladeren...</translation>
     </message>
     <message>
         <location line="+1"/>
         <source>Browse for start directory</source>
         <translation>Blader naar begin map</translation>
     </message>
     <message>
         <location line="+7"/>
diff --git a/libgui/languages/pt_BR.ts b/libgui/languages/pt_BR.ts
--- a/libgui/languages/pt_BR.ts
+++ b/libgui/languages/pt_BR.ts
@@ -383,17 +383,17 @@ para escrita: %2.</translation>
         <translation>&amp;Executar</translation>
     </message>
 </context>
 <context>
     <name>file_editor_tab</name>
     <message>
         <location filename="../src/m-editor/file-editor-tab.cc" line="+824"/>
         <source>Goto line</source>
-        <translation>Vá para linha...</translation>
+        <translation>Vá para linha</translation>
     </message>
     <message>
         <location line="+1"/>
         <source>Line number</source>
         <translation>Número da linha</translation>
     </message>
     <message>
         <location line="+108"/>
@@ -517,17 +517,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation>Buscar Arquivos...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Novo Arquivo</translation>
     </message>
     <message>
@@ -549,30 +549,35 @@ Do you want to choose another name?</sou
     <message>
         <location line="+2"/>
         <source>Go to current Octave directory</source>
         <translation>Vá para o diretório atual do Octave</translation>
     </message>
     <message>
         <location line="+5"/>
         <source>Set Octave directory</source>
-        <translation>Definir o diretório do Octave</translation>
+        <translation>Escolher o diretório do Octave</translation>
     </message>
     <message>
         <location line="+2"/>
         <source>Set Octave directroy to current browser directory</source>
-        <translation>Definir diretório do Octave igual ao diretório atual no navegador</translation>
+        <translation>Configurar o diretório do Octave igual ao diretório atual no navegador</translation>
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
-        <translation>Exibir diretório pessoal</translation>
+        <translation>Exibir Diretório Pessoal</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Buscar no Diretório...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Buscar Arquivos...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Novo Arquivo...</translation>
     </message>
@@ -624,17 +629,17 @@ Do you want to choose another name?</sou
     <message>
         <location line="+3"/>
         <source>Load Data</source>
         <translation>Carregar Dados</translation>
     </message>
     <message>
         <location line="+6"/>
         <source>Set Current Directory</source>
-        <translation>Definir Diretório Atual</translation>
+        <translation>Escolher Diretório Atual</translation>
     </message>
     <message>
         <location line="+8"/>
         <source>Rename</source>
         <translation>Renomear</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -674,17 +679,17 @@ para:</translation>
     <message>
         <location line="+14"/>
         <source>Can not delete a directory that is not empty</source>
         <translation>Não é possível excluir um diretório que não está vázio</translation>
     </message>
     <message>
         <location line="+144"/>
         <source>Set directory of file browser</source>
-        <translation>Definir diretório do Navegador de Arquivos</translation>
+        <translation>Escolher diretório do Navegador de Arquivos</translation>
     </message>
     <message>
         <location line="+29"/>
         <source>Create File</source>
         <translation>Criar Arquivo</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -738,19 +743,19 @@ para:</translation>
     <message>
         <location line="+7"/>
         <source>&lt;html&gt;&lt;body&gt;
 &lt;p&gt;We hope you find Octave to be a useful tool.&lt;/p&gt;
 &lt;p&gt;If you encounter problems, there are a number of ways to get help, including commercial support options, a mailing list, a wiki, and other commnity-based support channels.
 You can find more information about each of these by visiting &lt;a href=&quot;http://octave.org/support.html&quot;&gt;http://octave.org/support.html&lt;/a&gt; (opens in external browser).&lt;/p&gt;
 &lt;/body&gt;&lt;/html&gt;</source>
         <translation>&lt;html&gt;&lt;body&gt;
-&lt;p&gt;Esperamos que o Octave seja uma ferramenta últil.&lt;/p&gt;
-&lt;p&gt;Se encontrar problemas, existem várias maneiras de pedir ajuda incluindo suporte comercial, lista de email, wiki, e outros canais de suporte da comunidade.
-Você pode encontrar mais informações sobre cada um deles visitando &lt;a href=&quot;http://octave.org/support.html&quot;&gt;http://octave.org/support.html&lt;/a&gt; (opens in external browser).&lt;/p&gt;
+&lt;p&gt;Esperamos que você ache o Octave uma ferramenta útil.&lt;/p&gt;
+&lt;p&gt;Se encontrar dificuldades, existem várias formas de obter assistência, incluindo suporte comercial, listas de email, wiki, e outros canais de suporte comunitários.
+Você pode encontrar mais informações sobre cada um deles visitando &lt;a href=&quot;http://octave.org/support.html&quot;&gt;http://octave.org/support.html&lt;/a&gt; (link abre em navegador externo).&lt;/p&gt;
 &lt;/body&gt;&lt;/html&gt;</translation>
     </message>
     <message>
         <location line="+21"/>
         <source>&lt;html&gt;&lt;head&gt;
 &lt;style&gt;
 a:link { text-decoration: underline; color: #0000ff; }
 &lt;/style&gt;
@@ -826,32 +831,32 @@ a:link { text-decoration: underline; col
     <message>
         <location line="+1"/>
         <source>Replace &amp;All</source>
         <translation>Substituir &amp;Todos(as)</translation>
     </message>
     <message>
         <location line="+2"/>
         <source>&amp;More...</source>
-        <translation>&amp;Mais...</translation>
+        <translation>&amp;Mais opções...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>&amp;Whole words</source>
         <translation>Palavras &amp;inteiras</translation>
     </message>
     <message>
         <location line="+1"/>
         <source>Regular E&amp;xpressions</source>
         <translation>Expressões &amp;Regulares</translation>
     </message>
     <message>
         <location line="+1"/>
         <source>Search &amp;backward</source>
-        <translation>Buscar de trás para frente</translation>
+        <translation>Buscar de &amp;trás para frente</translation>
     </message>
     <message>
         <location line="+1"/>
         <source>Search se&amp;lection</source>
         <translation>Buscar dentro do texto &amp;selecionado</translation>
     </message>
     <message>
         <location line="+72"/>
@@ -1104,27 +1109,27 @@ a:link { text-decoration: underline; col
         <location line="+7"/>
         <source>&lt;html&gt;&lt;body&gt;
 &lt;p&gt;You seem to be using the Octave graphical interface for the first time on this computer.
 Click &apos;Next&apos; to create a configuration file and launch Octave.&lt;/p&gt;
 &lt;p&gt;The configuration file is stored in&lt;br&gt;%1.&lt;/p&gt;
 &lt;/body&gt;&lt;/html&gt;</source>
         <translation>&lt;html&gt;&lt;body&gt;
 &lt;p&gt;Parece que você está usando a interface gráfica pela primeira vez neste computador.
-Clique em próximo para criar um arquivo de configuração e lançar o Octave.&lt;/p&gt;
+Clique em &apos;Avançar&apos; para criar um arquivo de configuração e iniciar o Octave.&lt;/p&gt;
 &lt;p&gt;O arquivo de configuração é armazenado em&lt;br&gt;%1.&lt;/p&gt;
 &lt;/body&gt;&lt;/html&gt;</translation>
     </message>
 </context>
 <context>
     <name>main_window</name>
     <message>
         <location filename="../src/main-window.cc" line="-1758"/>
         <source>Load Workspace</source>
-        <translation>Carregar Ambiente de Trabalho...</translation>
+        <translation>Carregar Ambiente de Trabalho</translation>
     </message>
     <message>
         <location line="+678"/>
         <location line="+905"/>
         <source>About Octave</source>
         <translation>Sobre o Octave</translation>
     </message>
     <message>
@@ -1175,17 +1180,17 @@ Clique em próximo para criar um arquivo de configuração e lançar o Octave.&lt;/p&gt;
     <message>
         <location line="+6"/>
         <source>Paste</source>
         <translation>Colar</translation>
     </message>
     <message>
         <location line="-1327"/>
         <source>Save Workspace As</source>
-        <translation>Salvar Ambiente de Trabalho Como...</translation>
+        <translation>Salvar Ambiente de Trabalho Como</translation>
     </message>
     <message>
         <location line="+84"/>
         <source>The file %1
 can not be executed because its name
 is not a valid identifier.
 
 Do you want to execute
@@ -1281,17 +1286,17 @@ instead?</source>
     <message>
         <location line="-1311"/>
         <source>Octave</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+1160"/>
         <source>Load Workspace...</source>
-        <translation>Carregar ambiente de trabalho...</translation>
+        <translation>Carregar Ambiente de Trabalho...</translation>
     </message>
     <message>
         <location line="+3"/>
         <source>Save Workspace As...</source>
         <translation>Salvar Ambiente de Trabalho Como...</translation>
     </message>
     <message>
         <location line="+38"/>
@@ -1772,17 +1777,17 @@ Octave GUI precisa ser fechado agora.</t
     <message>
         <location line="+35"/>
         <source>Editor Styles</source>
         <translation>Apresentação</translation>
     </message>
     <message>
         <location line="+17"/>
         <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Select font, font size (as a difference from the default size), font style (&lt;b&gt;b&lt;/b&gt;old, &lt;b&gt;i&lt;/b&gt;talic, &lt;b&gt;u&lt;/b&gt;nderline), text color and background color (for the latter, the color pink (255,0,255) is a placeholder for the default background color).&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
-        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Selecionar fonte, tamanho da fonte (como uma diferença do tamanho padrão), estilo da fonte (&lt;b&gt;n&lt;/b&gt;egrito, &lt;b&gt;i&lt;/b&gt;tálico, &lt;b&gt;s&lt;/b&gt;ublinhado), cor de texto e cor de fundo (a cor rosa (255,0,255) é a cor de fundo padrão).&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
+        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Selecionar fonte, tamanho da fonte (variacão com relacão ao tamanho padrão), estilo da fonte (&lt;b&gt;n&lt;/b&gt;egrito, &lt;b&gt;i&lt;/b&gt;tálico, &lt;b&gt;s&lt;/b&gt;ublinhado), cor da fonte e cor de fundo (a cor rosa (255,0,255) é a cor de fundo padrão).&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
     </message>
     <message>
         <location line="+191"/>
         <source>Terminal Colors</source>
         <translation>Cores do Terminal</translation>
     </message>
     <message>
         <location line="-100"/>
diff --git a/libgui/languages/pt_PT.ts b/libgui/languages/pt_PT.ts
--- a/libgui/languages/pt_PT.ts
+++ b/libgui/languages/pt_PT.ts
@@ -1186,17 +1186,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation>Procurar Ficheiros...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Novo Ficheiro</translation>
     </message>
     <message>
@@ -1232,16 +1232,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Mostrar directório pessoal</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation>Procurar no directório...</translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Encontrar Ficheiros...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Novo Ficheiro...</translation>
     </message>
diff --git a/libgui/languages/ru_RU.ts b/libgui/languages/ru_RU.ts
--- a/libgui/languages/ru_RU.ts
+++ b/libgui/languages/ru_RU.ts
@@ -519,17 +519,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation>Найти файлы...</translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation>Новый файл</translation>
     </message>
     <message>
@@ -565,16 +565,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation>Показывать домашнюю директорию</translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation></translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation>Найти файлы...</translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation>Новый файл...</translation>
     </message>
diff --git a/libgui/languages/uk_UA.ts b/libgui/languages/uk_UA.ts
--- a/libgui/languages/uk_UA.ts
+++ b/libgui/languages/uk_UA.ts
@@ -1169,17 +1169,17 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+15"/>
         <source>Set Browser Directory...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+379"/>
-        <source>Find Files ...</source>
+        <source>Find Files...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+13"/>
         <source>New File</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
@@ -1215,16 +1215,21 @@ Do you want to choose another name?</sou
     </message>
     <message>
         <location line="+7"/>
         <source>Show Home Directory</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+16"/>
+        <source>Search Directory...</source>
+        <translation type="unfinished"></translation>
+    </message>
+    <message>
+        <location line="+4"/>
         <source>Find Files...</source>
         <translation type="unfinished"></translation>
     </message>
     <message>
         <location line="+4"/>
         <source>New File...</source>
         <translation type="unfinished"></translation>
     </message>
diff --git a/libgui/src/dialog.cc b/libgui/src/dialog.cc
--- a/libgui/src/dialog.cc
+++ b/libgui/src/dialog.cc
@@ -486,18 +486,30 @@ void FileDialog::accept (void)
   string_result = selectedFiles ();
 
   // Matlab expects just the file name, whereas the file dialog gave us
   // pull path names, so fix it.
 
   for (int i = 0; i < string_result.size (); i++)
     string_result[i] = QFileInfo (string_result[i]).fileName ();
 
+  path = directory ().absolutePath ();
 
-  path = directory ().absolutePath ();
+  // if not showing only dirs, add end slash for the path component
+  if (testOption (QFileDialog::ShowDirsOnly)  == false)
+    path = path + "/";
+  else
+    {
+      // if name was provided in uigetdir, add to path
+      if (string_result.size() > 0)
+        path = path + "/" + string_result[0];
+    }
+
+  // convert to native slashes
+  path = QDir::toNativeSeparators (path);
 
   QStringList name_filters = nameFilters ();
   idx = name_filters.indexOf (selectedNameFilter ()) + 1;
 
   // send the selected info
   emit finish_input (string_result, path, idx);
   done (QDialog::Accepted);
 }
diff --git a/libgui/src/files-dock-widget.cc b/libgui/src/files-dock-widget.cc
--- a/libgui/src/files-dock-widget.cc
+++ b/libgui/src/files-dock-widget.cc
@@ -472,22 +472,33 @@ files_dock_widget::headercontextmenu_req
 
 void
 files_dock_widget::contextmenu_requested (const QPoint& mpos)
 {
 
   QMenu menu (this);
 
   QModelIndex index = _file_tree_view->indexAt (mpos);
-  //QAbstractItemModel *m = _file_tree_view->model ();
 
   if (index.isValid ())
     {
       QFileInfo info = _file_system_model->fileInfo (index);
 
+      QItemSelectionModel *m = _file_tree_view->selectionModel ();
+      QModelIndexList sel = m->selectedRows ();
+
+      // check if item at mouse position is seleccted
+      if (! sel.contains (index))
+        { // is not selected -> clear actual selection and select this item
+          m->setCurrentIndex(index,
+                  QItemSelectionModel::Clear | QItemSelectionModel::Select |
+                  QItemSelectionModel::Rows);
+        }
+
+      // construct the context menu depending on item
       menu.addAction (QIcon (":/actions/icons/fileopen.png"), tr ("Open"),
                       this, SLOT (contextmenu_open (bool)));
 
       menu.addAction (tr ("Open in Default Application"),
                       this, SLOT (contextmenu_open_in_app (bool)));
 
       menu.addAction (tr ("Copy Selection to Clipboard"),
                       this, SLOT (contextmenu_copy_selection (bool)));
@@ -501,51 +512,60 @@ files_dock_widget::contextmenu_requested
 
       if (info.isDir ())
         {
           menu.addSeparator ();
           menu.addAction (QIcon (":/actions/icons/ok.png"),
                           tr ("Set Current Directory"),
                           this, SLOT (contextmenu_setcurrentdir (bool)));
           menu.addSeparator ();
-          menu.addAction (tr ("Find Files ..."), this,
+          menu.addAction (QIcon (":/actions/icons/findf.png"),
+                          tr ("Find Files..."), this,
                           SLOT (contextmenu_findfiles (bool)));
         }
 
       menu.addSeparator ();
-      menu.addAction (tr ("Rename"), this, SLOT (contextmenu_rename (bool)));
-      menu.addAction (QIcon (":/actions/icons/editdelete.png"), tr ("Delete"),
-                      this, SLOT (contextmenu_delete (bool)));
+      menu.addAction (tr ("Rename..."), this, SLOT (contextmenu_rename (bool)));
+      menu.addAction (QIcon (":/actions/icons/editdelete.png"),
+                      tr ("Delete..."), this, SLOT (contextmenu_delete (bool)));
 
       if (info.isDir ())
         {
           menu.addSeparator ();
           menu.addAction (QIcon (":/actions/icons/filenew.png"),
-                          tr ("New File"),
+                          tr ("New File..."),
                           this, SLOT (contextmenu_newfile (bool)));
           menu.addAction (QIcon (":/actions/icons/folder_new.png"),
-                          tr ("New Directory"),
+                          tr ("New Directory..."),
                           this, SLOT (contextmenu_newdir (bool)));
         }
 
+      // show the menu
       menu.exec (_file_tree_view->mapToGlobal (mpos));
 
     }
 }
 
 void
 files_dock_widget::contextmenu_open (bool)
 {
 
   QItemSelectionModel *m = _file_tree_view->selectionModel ();
   QModelIndexList rows = m->selectedRows ();
 
   for ( QModelIndexList::iterator it = rows.begin (); it != rows.end (); it++)
     {
-      item_double_clicked (*it);
+      QFileInfo file = _file_system_model->fileInfo (*it);
+      if (file.exists ())
+        {
+          if (file.isFile ())
+            emit open_file (file.absoluteFilePath ());
+          else
+            set_current_directory (file.absoluteFilePath ());
+        }
     }
 }
 
 void
 files_dock_widget::contextmenu_open_in_app (bool)
 {
   QItemSelectionModel *m = _file_tree_view->selectionModel ();
   QModelIndexList rows = m->selectedRows ();
diff --git a/libgui/src/m-editor/file-editor-interface.h b/libgui/src/m-editor/file-editor-interface.h
--- a/libgui/src/m-editor/file-editor-interface.h
+++ b/libgui/src/m-editor/file-editor-interface.h
@@ -56,16 +56,18 @@ public:
   virtual void
   handle_update_breakpoint_marker_request (bool insert, const QString& file,
                                            int line) = 0;
 
   virtual void handle_edit_file_request (const QString& file) = 0;
 
   virtual void set_focus () = 0;
 
+  virtual void empty_script (bool, bool) = 0;
+
 public slots:
   virtual void request_new_file (const QString& command = QString ()) = 0;
   virtual void request_new_script (const QString& command = QString ()) = 0;
   virtual void request_new_function (bool) = 0;
   virtual void request_open_file () = 0;
   virtual void request_open_file (const QString& openFileName, int line = -1,
                                   bool debug_pointer = false,
                                   bool breakpoint_marker = false,
diff --git a/libgui/src/m-editor/file-editor.cc b/libgui/src/m-editor/file-editor.cc
--- a/libgui/src/m-editor/file-editor.cc
+++ b/libgui/src/m-editor/file-editor.cc
@@ -31,16 +31,17 @@ along with Octave; see the file COPYING.
 #include <QVBoxLayout>
 #include <QApplication>
 #include <QFile>
 #include <QFont>
 #include <QFileDialog>
 #include <QMessageBox>
 #include <QStyle>
 #include <QTextStream>
+#include <QTabBar>
 #include <QProcess>
 #include <QInputDialog>
 
 #include "octave-link.h"
 #include "utils.h"
 
 file_editor::file_editor (QWidget *p)
   : file_editor_interface (p)
@@ -1542,10 +1543,72 @@ file_editor::check_actions ()
   _close_action->setEnabled (have_tabs);
   _close_all_action->setEnabled (have_tabs);
   _close_others_action->setEnabled (have_tabs && _tab_widget->count () > 1);
 
   _undo_action->setEnabled (have_tabs);
   _redo_action->setEnabled (have_tabs);
 }
 
+// empty_script determines whether we have to create an empty script
+// 1. At startup, when the editor has to be (really) visible
+//    (Here we can not use the visibility changed signal)
+// 2. When the editor becomes visible when octave is running
+void
+file_editor::empty_script (bool startup, bool visible)
+{
+  bool real_visible;
+
+  if (startup)
+    real_visible = isVisible ();
+  else
+    real_visible = visible;
+
+  if (! real_visible || _tab_widget->count () > 0)
+    return;
+
+  if (startup && ! isFloating ())
+    {
+      // check is editor is really visible or hidden between tabbed widgets
+      QList<QTabBar *> tab_list = main_win ()->findChildren<QTabBar *>();
+
+      bool in_tab = false;
+      int i = 0;
+      while ((i < tab_list.count ()) && (! in_tab))
+        {
+          QTabBar *tab = tab_list.at (i);
+          i++;
+
+          int j = 0;
+          while ((j < tab->count ()) && (! in_tab))
+            {
+              // check all tabs for the editor
+              if (tab->tabText (j) == windowTitle ())
+                {
+                  // editor is in this tab widget
+                  in_tab = true;
+                  int top = tab->currentIndex ();
+                  if (top > -1 && tab->tabText (top) == windowTitle ())
+                    real_visible = true;  // and is the current tab
+                  else
+                    return; // not current tab -> not visible
+                }
+              j++;
+            }
+        }
+    }
+
+  request_new_file ("");
+}
+
+// This slot is a reimplementation of the virtual slot in octave_dock_widget.
+// We need this for creating an empty script when the editor has no open files
+// and is made visible
+void
+file_editor::handle_visibility (bool visible)
+  {
+    empty_script (false, visible);
+
+    if (visible && ! isFloating ())
+      focus ();
+  }
 
 #endif
diff --git a/libgui/src/m-editor/file-editor.h b/libgui/src/m-editor/file-editor.h
--- a/libgui/src/m-editor/file-editor.h
+++ b/libgui/src/m-editor/file-editor.h
@@ -53,16 +53,18 @@ public:
   QMenu *debug_menu (void);
   QToolBar *toolbar (void);
 
   void set_focus (void);
   void handle_enter_debug_mode (void);
   void handle_exit_debug_mode (void);
 
   void check_actions (void);
+  void empty_script (bool startup, bool visible);
+
 signals:
 
   void fetab_settings_changed (const QSettings *settings);
   void fetab_close_request (const QWidget* ID, bool app_closing = false);
   void fetab_change_request (const QWidget* ID);
   void fetab_file_name_query (const QWidget* ID);
   // Save is a ping-pong type of communication
   void fetab_save_file (const QWidget* ID, const QString& fileName,
@@ -167,16 +169,18 @@ public slots:
   void handle_edit_file_request (const QString& file);
 
   // Tells the editor to react on changed settings.
   void notice_settings (const QSettings *settings);
 
   // Tells the ditor to dis- or enable some shortcuts
   void set_shortcuts (bool set_shortcuts);
 
+  void handle_visibility (bool visible);
+
 
 protected slots:
   void copyClipboard ();
   void pasteClipboard ();
 
 private slots:
 
   void request_open_file (const QString& fileName, int line = -1,
diff --git a/libgui/src/main-window.cc b/libgui/src/main-window.cc
--- a/libgui/src/main-window.cc
+++ b/libgui/src/main-window.cc
@@ -1005,16 +1005,19 @@ main_window::write_settings (void)
 
 // Connecting the signals emitted when the visibility of a widget changes.
 // This has to be done after the window is shown (see octave-gui.cc)
 void
 main_window::connect_visibility_changed (void)
 {
   foreach (octave_dock_widget *widget, dock_widget_list ())
     widget->connect_visibility_changed ();
+
+  // Main window completely shown, determine whether to create an empty script
+  editor_window->empty_script (true, false);
 }
 
 void
 main_window::copyClipboard (void)
 {
   if (_current_directory_combo_box->hasFocus ())
     {
       QLineEdit * edit = _current_directory_combo_box->lineEdit ();
diff --git a/libinterp/corefcn/ls-mat4.cc b/libinterp/corefcn/ls-mat4.cc
--- a/libinterp/corefcn/ls-mat4.cc
+++ b/libinterp/corefcn/ls-mat4.cc
@@ -580,10 +580,10 @@ save_mat_binary_data (std::ostream& os, 
       Matrix m = ::real (m_cmplx);
       os.write (reinterpret_cast<const char *> (m.data ()), 8 * len);
       m = ::imag (m_cmplx);
       os.write (reinterpret_cast<const char *> (m.data ()), 8 * len);
     }
   else
     gripe_wrong_type_arg ("save", tc, false);
 
-  return os;
+  return ! os.fail ();
 }
diff --git a/libinterp/corefcn/urlwrite.cc b/libinterp/corefcn/urlwrite.cc
--- a/libinterp/corefcn/urlwrite.cc
+++ b/libinterp/corefcn/urlwrite.cc
@@ -413,17 +413,17 @@ urlwrite (\"http://www.google.com/search
     }
 
   unwind_protect_safe frame;
 
   frame.add_fcn (delete_file, filename);
 
   url_transfer curl = url_transfer (url, ofile);
 
-  if (! curl.is_valid ())
+  if (curl.is_valid ())
     {
       curl.http_action (param, method);
 
       ofile.close ();
 
       if (curl.good ())
         frame.discard ();
 
diff --git a/libinterp/octave-value/ov-cell.cc b/libinterp/octave-value/ov-cell.cc
--- a/libinterp/octave-value/ov-cell.cc
+++ b/libinterp/octave-value/ov-cell.cc
@@ -775,17 +775,17 @@ octave_cell::save_ascii (std::ostream& o
       for (octave_idx_type i = 0; i < d.numel (); i++)
         {
           octave_value o_val = tmp.elem (i);
 
           // Recurse to print sub-value.
           bool b = save_ascii_data (os, o_val, CELL_ELT_TAG, false, 0);
 
           if (! b)
-            return os;
+            return ! os.fail ();
         }
     }
   else
     {
       // Keep this case, rather than use generic code above for backward
       // compatiability. Makes load_ascii much more complex!!
       os << "# rows: " << rows () << "\n"
          << "# columns: " << columns () << "\n";
@@ -797,17 +797,17 @@ octave_cell::save_ascii (std::ostream& o
           for (octave_idx_type i = 0; i < tmp.rows (); i++)
             {
               octave_value o_val = tmp.elem (i, j);
 
               // Recurse to print sub-value.
               bool b = save_ascii_data (os, o_val, CELL_ELT_TAG, false, 0);
 
               if (! b)
-                return os;
+                return ! os.fail ();
             }
 
           os << "\n";
         }
     }
 
   return true;
 }
diff --git a/libinterp/octave-value/ov-class.cc b/libinterp/octave-value/ov-class.cc
--- a/libinterp/octave-value/ov-class.cc
+++ b/libinterp/octave-value/ov-class.cc
@@ -1246,17 +1246,17 @@ octave_class::save_ascii (std::ostream& 
   octave_map::iterator i = m.begin ();
   while (i != m.end ())
     {
       octave_value val = map.contents (i);
 
       bool b = save_ascii_data (os, val, m.key (i), false, 0);
 
       if (! b)
-        return os;
+        return ! os.fail ();
 
       i++;
     }
 
   return true;
 }
 
 bool
@@ -1375,17 +1375,17 @@ octave_class::save_binary (std::ostream&
   octave_map::iterator i = m.begin ();
   while (i != m.end ())
     {
       octave_value val = map.contents (i);
 
       bool b = save_binary_data (os, val, m.key (i), "", 0, save_as_floats);
 
       if (! b)
-        return os;
+        return ! os.fail ();
 
       i++;
     }
 
   return true;
 }
 
 bool
diff --git a/libinterp/octave-value/ov-fcn-handle.cc b/libinterp/octave-value/ov-fcn-handle.cc
--- a/libinterp/octave-value/ov-fcn-handle.cc
+++ b/libinterp/octave-value/ov-fcn-handle.cc
@@ -375,17 +375,17 @@ octave_fcn_handle::save_ascii (std::ostr
       if (varlen > 0)
         {
           os << "# length: " << varlen << "\n";
 
           for (std::list<symbol_table::symbol_record>::const_iterator
                p = vars.begin (); p != vars.end (); p++)
             {
               if (! save_ascii_data (os, p->varval (0), p->name (), false, 0))
-                return os;
+                return ! os.fail ();
             }
         }
     }
   else
     {
       octave_function *f = function_value ();
       std::string fnm = f ? f->fcn_file_name () : std::string ();
 
@@ -551,17 +551,17 @@ octave_fcn_handle::save_binary (std::ost
 
       if (varlen > 0)
         {
           for (std::list<symbol_table::symbol_record>::const_iterator
                p = vars.begin (); p != vars.end (); p++)
             {
               if (! save_binary_data (os, p->varval (0), p->name (),
                                       "", 0, save_as_floats))
-                return os;
+                return ! os.fail ();
             }
         }
     }
   else
     {
       std::ostringstream nmbuf;
 
       octave_function *f = function_value ();
diff --git a/libinterp/octave-value/ov-struct.cc b/libinterp/octave-value/ov-struct.cc
--- a/libinterp/octave-value/ov-struct.cc
+++ b/libinterp/octave-value/ov-struct.cc
@@ -764,17 +764,17 @@ octave_struct::save_ascii (std::ostream&
     {
       std::string key = keys(i);
 
       octave_value val = map.contents (key);
 
       bool b = save_ascii_data (os, val, key, false, 0);
 
       if (! b)
-        return os;
+        return ! os.fail ();
     }
 
   return true;
 }
 
 bool
 octave_struct::load_ascii (std::istream& is)
 {
@@ -889,17 +889,17 @@ octave_struct::save_binary (std::ostream
     {
       std::string key = keys(i);
 
       octave_value val = map.contents (key);
 
       bool b = save_binary_data (os, val, key, "", 0, save_as_floats);
 
       if (! b)
-        return os;
+        return ! os.fail ();
     }
 
   return true;
 }
 
 bool
 octave_struct::load_binary (std::istream& is, bool swap,
                             oct_mach_info::float_format fmt)
@@ -1466,17 +1466,17 @@ octave_scalar_struct::save_ascii (std::o
     {
       std::string key = keys(i);
 
       octave_value val = map.contents (key);
 
       bool b = save_ascii_data (os, val, key, false, 0);
 
       if (! b)
-        return os;
+        return ! os.fail ();
     }
 
   return true;
 }
 
 bool
 octave_scalar_struct::load_ascii (std::istream& is)
 {
@@ -1550,17 +1550,17 @@ octave_scalar_struct::save_binary (std::
     {
       std::string key = keys(i);
 
       octave_value val = map.contents (key);
 
       bool b = save_binary_data (os, val, key, "", 0, save_as_floats);
 
       if (! b)
-        return os;
+        return ! os.fail ();
     }
 
   return true;
 }
 
 bool
 octave_scalar_struct::load_binary (std::istream& is, bool swap,
                                    oct_mach_info::float_format fmt)
diff --git a/scripts/help/__makeinfo__.m b/scripts/help/__makeinfo__.m
--- a/scripts/help/__makeinfo__.m
+++ b/scripts/help/__makeinfo__.m
@@ -124,20 +124,20 @@ function [retval, status] = __makeinfo__
       error ("__makeinfo__: could not create temporary file");
     endif
     fwrite (fid, text);
     fclose (fid);
 
     ## Take action depending on output type
     switch (lower (output_type))
       case "plain text"
-        cmd = sprintf ("%s --no-headers --no-warn --force --no-validate %s",
+        cmd = sprintf ("%s --no-headers --no-warn --force --no-validate --output=- %s",
                        makeinfo_program (), name);
       case "html"
-        cmd = sprintf ("%s --no-headers --html --no-warn --no-validate --force %s",
+        cmd = sprintf ("%s --no-headers --html --no-warn --no-validate --force --output=- %s",
                        makeinfo_program (), name);
       otherwise
         error ("__makeinfo__: unsupported output type: '%s'", output_type);
     endswitch
 
     ## Call makeinfo
     [status, retval] = system (cmd);
 
diff --git a/scripts/miscellaneous/debug.m b/scripts/miscellaneous/debug.m
--- a/scripts/miscellaneous/debug.m
+++ b/scripts/miscellaneous/debug.m
@@ -32,19 +32,23 @@
 ##
 ## @item dbstatus
 ## List all breakpoints.
 ##
 ## @item dbwhere
 ## Report the current file and line number where execution is stopped.
 ##
 ## @item dbtype
-## List the function where execution is currently stopped, enumerating
+## Display the code of the function being debugged, enumerating
 ## the line numbers.
 ##
+## @item dblist
+## List 10 lines of code centered around the line number where execution is
+## stopped.
+##
 ## @item  dbstep
 ## @itemx dbnext
 ## Execute (step) one or more lines, follow execution into (step into) a
 ## function call, or execute until the end of a function (step out), and
 ## re-enter debug mode.
 ##
 ## @item dbcont
 ## Continue normal code execution from the debug prompt.
diff --git a/scripts/plot/util/graphics_toolkit.m b/scripts/plot/util/graphics_toolkit.m
--- a/scripts/plot/util/graphics_toolkit.m
+++ b/scripts/plot/util/graphics_toolkit.m
@@ -61,16 +61,20 @@ function retval = graphics_toolkit (name
     [hlist, name] = deal (name, hlist);
     if (! all (isfigure (hlist)))
       error ("graphics_toolkit: invalid figure handle list HLIST");
     elseif (! ischar (name))
       error ("graphics_toolkit: invalid graphics toolkit NAME");
     endif
   endif
 
+  if (! any (strcmp (available_graphics_toolkits (), name)))
+    error ("graphics_toolkit: %s toolkit is not available", name);
+  endif
+
   if (! any (strcmp (loaded_graphics_toolkits (), name)))
     feval (["__init_", name, "__"]);
     if (! any (strcmp (loaded_graphics_toolkits (), name)))
       error ("graphics_toolkit: %s toolkit was not correctly loaded", name);
     endif
   endif
 
   if (isempty (hlist))
diff --git a/scripts/testfun/assert.m b/scripts/testfun/assert.m
--- a/scripts/testfun/assert.m
+++ b/scripts/testfun/assert.m
@@ -51,17 +51,17 @@
 ## its dimensions must agree with those of @var{observed} and @var{expected}
 ## and tests are performed on an element-wise basis.
 ## @end table
 ## @seealso{test, fail, error}
 ## @end deftypefn
 
 function assert (cond, varargin)
 
-  if (nargin == 0 || nargin > 3)
+  if (nargin == 0)
     print_usage ();
   endif
 
   persistent call_depth = -1;
   persistent errmsg;
 
   call_depth++;
 
@@ -79,18 +79,20 @@ function assert (cond, varargin)
       else
         error (varargin{:});
       endif
     endif
   else
     expected = varargin{1};
     if (nargin < 3)
       tol = 0;
+    elseif (nargin == 3)
+      tol = varargin{2};
     else
-      tol = varargin{2};
+      print_usage ();
     endif
 
     ## Add to list as the errors accumulate.  If empty at end then no errors.
     err.index = {};
     err.observed = {};
     err.expected = {};
     err.reason = {};
 
@@ -391,16 +393,18 @@ endfunction
 %!assert (1)
 %!error assert (0)
 %!assert (ones (3,1))
 %!assert (ones (1,3))
 %!assert (ones (3,4))
 %!error assert ([1,0,1])
 %!error assert ([1;1;0])
 %!error assert ([1,0;1,1])
+%!error <2-part error> assert (false, "%s %s", "2-part", "error")
+%!error <2-part error> assert (false, "TST:msg_id", "%s %s", "2-part", "error")
 
 ## scalars
 %!error <Dimensions don't match> assert (3, [3,3])
 %!error <Dimensions don't match> assert (3, [3,3; 3,3])
 %!error <Dimensions don't match> assert ([3,3; 3,3], 3)
 %!assert (3, 3)
 %!error <Abs err 1 exceeds tol> assert (3, 4)
 %!assert (3+eps, 3, eps)
diff --git a/src/shared-fcns.h b/src/shared-fcns.h
--- a/src/shared-fcns.h
+++ b/src/shared-fcns.h
@@ -18,16 +18,17 @@ You should have received a copy of the G
 along with Octave; see the file COPYING.  If not, see
 <http://www.gnu.org/licenses/>.
 
 */
 
 #if defined (__WIN32__) && ! defined (_POSIX_VERSION)
 
 #include <windows.h>
+#include <tlhelp32.h>
 
 #ifdef _MSC_VER
 #define popen _popen
 #define pclose _pclose
 #endif
 
 static std::string
 w32_get_octave_home (void)
